columnflow/columnar_util.py

```diff
@@ -1952,6 +1974,140 @@ def get_next(self) -> tuple | None:
+            - guard each read operation of a partition by locks
+            - add materialized partitions that might overlap with another chunk in a temporary cache
+            -
+        """
+        # fill the chunk -> partitions mapping once
+        with self.chunk_to_partitions_lock:
+            if not self.chunk_to_partitions:
+                self.chunk_to_partitions.clear()
+                divs = self.dak_array.divisions
+                # note: a hare-and-tortoise algorithm could be possible to get the mapping with less
+                # than n^2 complexity, but for our case with ~30 chunks this should be ok (for now)
+                for _chunk_index in range(int(math.ceil(len(self) / max_chunk_size))):
+                    _entry_start = _chunk_index * max_chunk_size
+                    _entry_stop = min(_entry_start + max_chunk_size, len(self))
+                    partitions = []
+                    for p, (p_start, p_stop) in enumerate(zip(divs[:-1], divs[1:])):
+                        if p_stop <= _entry_start:
+                            continue
+                        if p_start >= _entry_stop:
+                            break
+                        partitions.append(p)
+                        self.partition_cache[p].chunks.append(_chunk_index)
+                    self.chunk_to_partitions[_chunk_index] = partitions
+
+        # read partitions one at a time and store parts that make up the chunk for concatenation
+        parts = []
+        for p in self.chunk_to_partitions[chunk_index]:
+            # obtain the array
+            with self.partition_locks[p]:
+                # remove this chunk for the list of chunks to be handled
+                self.partition_cache[p].chunks.remove(chunk_index)
+
+                if self.partition_cache[p].array is None:
+                    arr = self.dak_array.partitions[p].compute()
+                    # add to cache when there is a chunk left that will need it
+                    if self.partition_cache[p].chunks:
+                        self.partition_cache[p].array = arr
+                else:
+                    arr = self.partition_cache[p].array
+                    # remove from cache when there is no chunk left that would need it
+                    if not self.partition_cache[p].chunks:
+                        self.partition_cache[p].array = None
+
+            # add part for concatenation using entry info
+            div_start, div_stop = self.dak_array.divisions[p:p + 2]
+            part_start = max(entry_start - div_start, 0)
+            part_stop = min(entry_stop - div_start, div_stop - div_start)
+            parts.append(arr[part_start:part_stop])
+            # workaround for https://github.com/dask-contrib/dask-awkward/issues/140
+            # make the array non-optional, assuming it is not meant to be optional
+            # TODO: remove this workaround
+            parts[-1] = ak.fill_none(parts[-1], 0)
```

Comment:
Yeah, I'm also not happy with this line, but it's the most generic way I think.

(before going into detail, I think we are never saving optional types, that would end up in `Nones` in the first place, so consistency-wise we're safe)

The columns in those arrays can have arbitrary types, so using e.g. `-99999.0` would not play nicely with bool-types. On the other hand, `0` casts into every type we use, which is why I thought it's a natural choice (and it indeed does not lead to crashes :) ).

Reply:
Hm, ok. But just to make sure: what does such an array look like? I.e., does it have fields like `Jet.eta`? If no, I agree with you and see no problem in using `0` as default. If it does have such fields, maybe an error or at least a warning might be better imho

Reply:
Such an array would look like

```
array
├─ Jet
│  ├─ pt: [[1, 2], [None], [3, 4, 5], ...]
│  ├─ mass: [[6, 7], [None], [8, 9, 10], ...]
│  └─ ...
└─ ...
```

An in this case, the `None`'s would be filled with zeros. I agree that we should have a warning at this point and I'm going to add that. During production, this should be never raised I think (and hope ^^), since `None`'s are not written, so that `ak.fill_none` is effectively just a cast from (e.g.) `?float32` to `float32`.

Reply:
Ok, better approach: there is a `ak.drop_none` function now, which just removes `None`'s, but does not fill them up with a value that needs to be defined. I think this is the safest way of handling this right now 👍

Reply:
Awesome! I agree, this seems to be the safest option :+1: